### PR TITLE
Enhance param validation on refreshToken

### LINKED ISSUE #27: Uncaught TypeError: Argument 1 passed to EED_SquareOnsiteOAuth::encryptString() must be of the type string
Notice: Trying to get property 'nonce' of non-object in /dom16100/wp-content/plugins/eea-square-gateway/modules/EED_SquareOnsiteOAuth.module.php on line 557

Notice: Trying to get property 'application_id' of non-object in /dom16100/wp-content/plugins/eea-square-gateway/modules/EED_SquareOnsiteOAuth.module.php on line 580

Notice: Trying to get property 'access_token' of non-object in /dom16100/wp-content/plugins/eea-square-gateway/modules/EED_SquareOnsiteOAuth.module.php on line 584

Fatal error: Uncaught TypeError: Argument 1 passed to EED_SquareOnsiteOAuth::encryptString() must be of the type string, null given, called in /dom16100/wp-content/plugins/eea-square-gateway/modules/EED_SquareOnsiteOAuth.module.php on line 584 and defined in /dom16100/wp-content/plugins/eea-square-gateway/modules/EED_SquareOnsiteOAuth.module.php:793 Stack trace: #0 /dom16100/wp-content/plugins/eea-square-gateway/modules/EED_SquareOnsiteOAuth.module.php(584): EED_SquareOnsiteOAuth::encryptString(NULL, false) #1 /dom16100/wp-content/plugins/eea-square-gateway/modules/EED_SquareOnsiteOAuth.module.php(743): EED_SquareOnsiteOAuth::refreshToken(Object(EE_Payment_Method)) #2 /dom16100/wp-content/plugins/eea-square-gateway/payment_methods/SquareOnsite/forms/SettingsForm.php(113): EED_SquareOnsiteOAuth::checkAndRefreshToken(Object(EE_Payment_Method)) #3 /dom16100/wp-content/plugins/eea-square-gateway/payment_methods/SquareOnsite/forms/SettingsForm.php(64): EventEspresso\Square\payment_methods\SquareOnsite\forms\SettingsForm->valid in /dom16100/wp-content/plugins/eea-square-gateway/modules/EED_SquareOnsiteOAuth.module.php on line 793

‹ lorenzochrome21 — Event Smart 2022-01-27 at 3 09 44 PM

**Steps to reproduce**

Not sure. However, I do remember what I did on the site.

In this issue (https://github.com/eventespresso/eventsmart.com-website/issues/886#issuecomment-1018780369), I noticed that two sites deactivated the Square payment method. On this new site where I reported this issue, I followed these steps:

- Last Friday, Jan 21, I set up a new testing site through Chrome.
- I purchased the Square payment gateway to get access to the upgrade.
- I activated the Square in the payment methods screen.
- Then connected it in live mode.
- After, I set a Slack reminder to check to see if it was deactivated the following Monday.
- On Monday's check, the Square payment method was not deactivated.
- Today is Thursday so almost a week later, I went back to the site to check on the deactivated issue again. This time, I noticed a fatal error.

**Reported or seen on**

https://lorenzochrome21.makeeventsnotwar.com/wp-admin/admin.php?page=espresso_payment_settings#espresso_squareonsite_payment_settings

**Other notes**

I saw the same error on one of my testing sites (https://enzo12staging.wpengine.com/wp-admin/admin.php?page=espresso_payment_settings#espresso_squareonsite_payment_settings):

An error of type E_ERROR was caused in line 793 of the file /nas/content/live/enzo12staging/wp-content/plugins/eea-square-gateway/modules/EED_SquareOnsiteOAuth.module.php. Error message: Uncaught TypeError: Argument 1 passed to EED_SquareOnsiteOAuth::encryptString() must be of the type string, null given, called in /nas/content/live/enzo12staging/wp-content/plugins/eea-square-gateway/modules/EED_SquareOnsiteOAuth.module.php on line 584 and defined in /nas/content/live/enzo12staging/wp-content/plugins/eea-square-gateway/modules/EED_SquareOnsiteOAuth.module.php:793
Stack trace:
#0 /nas/content/live/enzo12staging/wp-content/plugins/eea-square-gateway/modules/EED_SquareOnsiteOAuth.module.php(584): EED_SquareOnsiteOAuth::encryptString(NULL, false)
#1 /nas/content/live/enzo12staging/wp-content/plugins/eea-square-gateway/modules/EED_SquareOnsiteOAuth.module.php(743): EED_SquareOnsiteOAuth::refreshToken(Object(EE_Payment_Method))
#2 /nas/content/live/enzo12staging/wp-content/plugins/eea-square-gateway/payment_methods/SquareOnsite/forms/SettingsForm.php(113): EED_SquareOnsiteOAuth::checkAndRefreshToken(Object(EE_Payment_Method))
#3 /nas/content/live/enzo12staging/wp-content/plugins/eea-square-gateway/

‹ Lorenzo Caum Site — WordPress 2022-01-27 at 3 30 06 PM

Tony shared this feedback in Slack support-desk:

> It means the keys have been set nothing and the pm is trying to use them. So the problem is not the fatal itself.... its that the keys are being blanked

### PR DESCRIPTION
**Pull request by knazart**
Resolves #27

## Problem this Pull Request solves
There was a fatal if the `refreshToken` action had missing parameters from an API `refresh` response. Work here prevents that from happening. 

## How has this been tested
Checking the logs. Refresh occurs every 2 days.

### CODE REVIEW

File: modules/EED_SquareOnsiteOAuth.module.php [hunk 1]
```diff
@@ -837,15 +840,12 @@ public static function decryptString(string $text, bool $sandbox_mode): string
...
     {
-        $sensitive_data = [
-            'access_token'  => '',
-            'refresh_token' => '',
-            'nonce'         => '',
-        ];
+        $sensitive_data = ['access_token', 'refresh_token', 'nonce'];
         foreach ($data as $key => $value) {
             $data[ $key ] = is_array($value) ? EED_SquareOnsiteOAuth::cleanDataArray($value) : $value;
+            $data[ $key ] = in_array($key, $sensitive_data) ? '***' : $value;
```
Comment (tn3rb) [inline modules/EED_SquareOnsiteOAuth.module.php:846]: this second line is overwriting the previous one and therefore negating what it has done

what about something like:

```suggestion
            $value = is_array($value) ? EED_SquareOnsiteOAuth::cleanDataArray($value) : $value;
            $data[ $key ] = in_array($key, $sensitive_data) ? '***' : $value;
```

**Review by tn3rb — COMMENTED** [review]
may have found an issue with some variable assignments 
so please have a look and confirm

**Review by tn3rb — APPROVED** [review]
cool 😎

**Comment by lorenzocaum** [comment]
Hey Nazar, can we start on testing with this pull request?

**Comment by lorenzocaum** [comment]
Testing sites with Event Espresso 4.10.26.p and the 27/param-validation branch.

- https://enzo12staging.wpengine.com/wp-admin/admin.php?page=espresso_payment_settings
- https://playground.eventespresso.com/wp-admin/admin.php?page=espresso_payment_settings

I connected both through live mode.

I set a Slack reminder to check back on Wednesday afternoon.

**Comment by knazart** [comment]
There should be a log `Refresh the token` in the Payment methods section. It logs every time the token is refreshed.

**Comment by lorenzocaum** [comment]
Hey Nazar, thanks for the tip. I see a couple of entries on the first site from today and both are for domain status verified. The second site doesn't have anything for today.

**Comment by lorenzocaum** [comment]
Hey Nazar, I rechecked both sites again today.

Both are still connected to Square and there isn't any fatal errors.

I checked the payment logs about the entry that should be made and I don't see any mentioning a refresh.

On this site (https://enzo12staging.wpengine.com), there are entries for today and yesterday. They all show this with the verified:

‹ Lorenzo Caum Site — WordPress 2022-02-03 at 3 21 57 PM

The other site doesn't have any entries for this month. Only some from end of January. (I think it may be because its an internal site and doesn't get traffic to trigger the cron.)

Do we have enough info to move forward with merging this in?

**Comment by lorenzocaum** [comment]
I checked these sites again:

- https://enzo12staging.wpengine.com/wp-admin/admin.php?page=espresso_payment_settings
- https://playground.eventespresso.com/wp-admin/admin.php?page=espresso_payment_settings

Both showed this:

‹ Lorenzo Caum Site — WordPress 2022-02-04 at 2 38 28 PM

I have been testing Square on MakeEventsNotWar so I'm not sure if that caused a problem since I'm using the same account.

**Comment by knazart** [comment]
Lorenzo Hi,

You need to check the logs. Or please give me access to those sites.
But yes, it can cause issues if you use the same test account at the same time.

**Comment by lorenzocaum** [comment]
Hi Nazar, I invited you to the sites below through your email (nazar@eventespresso.com):

- https://enzo12staging.wpengine.com/wp-admin/admin.php?page=espresso_payment_settings
- https://playground.eventespresso.com/wp-admin/admin.php?page=espresso_payment_settings

Let me know if you need anything else to look at the payment logs: 

https://github.com/eventespresso/eea-square-gateway/pull/28#issuecomment-1029369375

**Comment by knazart** [comment]
Well, the logs show that most are "This request could not be authorized."

But no info on Why... I do suspect the usage of the same account as the cause. Otherwise I don't know why those are authorized.

**Comment by lorenzocaum** [comment]
Hi Nazar, I'm using the same Square account for testing on ES staging. I think that may have caused the not authorized.

Can we merge this in and bring it over to ES staging so I can test there?

**Pull request merged**